### PR TITLE
Add an .mli interface for xapi_ha.ml

### PR DESCRIPTION
**Pull request by gaborigloi**
This mli file is useful, as it will hide quite a few functions in the .ml file.

### CODE REVIEW

**Comment by gaborigloi** [comment]
Feel free to squash when merging.

File: ocaml/xapi/xapi_ha.mli [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+(** Functions for implementing 'High Availability' (HA).
+    @group High Availability (HA) *)
+
+val ha_redo_log : Redo_log.redo_log
+(** The redo log instance used for HA *)
```
Comment (Frezzle) [inline ocaml/xapi/xapi_ha.mli]: The function comment doesn't need to be above the function declaration?

Comment (mseri) [inline ocaml/xapi/xapi_ha.mli]: I think the docstrings are fine either way

Comment (mseri) [inline ocaml/xapi/xapi_ha.mli]: For reference: http://caml.inria.fr/pub/docs/manual-ocaml/ocamldoc.html#sec333

**Comment by gaborigloi** [comment]
I thought it would be best to remove them, because keeping them both would introduce duplication, and the two comments could get out of sync. (IMO ideally we would have an Eclipse-like IDE that can show the ocamldoc for each value :smile: , like for Java )

File: ocaml/xapi/xapi_ha.mli (hunk 1 above)
Comment (gaborigloi) [inline ocaml/xapi/xapi_ha.mli]: I think it's safer to put it below the values, as the first comment is always associated with the module apparently, see http://caml.inria.fr/pub/docs/manual-ocaml/ocamldoc.html ("15.2.1  Placement of documentation comments").

**Comment by Frezzle** [comment]
If we are avoiding duplication then would that mean we would have some of the functions documented in the mli file and all the others will be documented in the ml file? In that case, it would be best to keep them all in one place for convenience (the ml file)?

**Comment by gaborigloi** [comment]
@Frezzle You are right, according to this we can keep the OCamldoc comments in the `.ml` file, and they will still be used by ocamldoc: http://caml.inria.fr/pub/docs/manual-ocaml/ocamldoc.html#s%3Aocamldoc-merge , and that will help @mseri too :smile: , so this solution would suit everyone hopefully, good idea!

**Comment by Frezzle** [comment]
This part is concerning?: `Only elements (values, types, classes, ...) declared in the .mli file are kept. In other terms, definitions from the .ml file that are not exported in the .mli file are not documented.`

Would this mean no documentation is generated for everything not in the mli file, meaning we wouldn't see it with ocp-* dev tools?...or is that completely separate?

**Comment by gaborigloi** [comment]
Indeed, but this doesn't really matter right now, because for most of our modules I couldn't generate documentation anyway, because it's not specified as part of a library in the `_oasis` file, see https://github.com/gaborigloi/xen-api/tree/ocamldoc and http://www.uk.xensource.com/~gabori/xapi_ocamldoc/ocamldoc/ generated from it.

Also, the functions hidden by the `.mli` are really an implementation detail of that module in my opinion, so users of that module won't really care about those, it would just create clutter. But yeah, we can tell ocamldoc to generate doc for the hidden values too.

The sections I referred to is this: `If a description is present in the .ml file and not in the .mli file, the .ml description is kept. In either case, all the information given in the .mli file is kept.`, I think this means that we can keep the ocamldoc comments in the `.ml` file.

**Comment by mseri** [comment]
@gaborigloi with my great sadness, I believe it is more consistent to keep them in the `mli` file if the choice is to avoid duplicating them.

@Frezzle yes, but that's expected. The mli gives you the _visible_ interface, so that is the one that is documented outside the package

**Comment by gaborigloi** [comment]
Oh, so then it's good as it is?

**Comment by mseri** [comment]
IMHO Yes

**Pull request merged**